File: tools/assignment_utils.py
"""
Shared utilities for assignment extraction scripts
Contains common functions, logging setup, and shared constants
"""

import os
import json
from datetime import datetime, timedelta
from dateutil import parser as dateutil_parser

# --- Constants for File Paths ---

# Directories
DATA_DIR = 'data'
SECRETS_DIR = 'secrets'

# Log file
LOGFILE = 'assignment_extraction.log'

# Data files
ALL_ASSIGNMENTS_FILE = os.path.join(DATA_DIR, 'all_assignments.json')
ALL_GCLASSROOM_ASSIGNMENTS_FILE = os.path.join(DATA_DIR, 'all_gclassroom_assignments.json')
CLASSROOM_ACCOUNTS_FILE = os.path.join(DATA_DIR, 'classroom_accounts.json')
JUPITER_ASSIGNMENTS_FILE = os.path.join(DATA_DIR, 'jupiter_assignments.json')
JUPITER_CLASSES_FILE = os.path.join(DATA_DIR, 'jupiter_classes.json')
COURSE_COLORS_FILE = os.path.join(DATA_DIR, 'course_colors.json')
CLASS_COLORS_FILE = os.path.join(DATA_DIR, 'class_colors.json')

# Secret files
JUPITER_SECRET_FILE = os.path.join(SECRETS_DIR, 'jupiter_secret.json')
GOOGLE_CLASSROOM_COOKIES_FILE = os.path.join(SECRETS_DIR, 'google_classroom_cookies.pkl')
BACKUP_COOKIES_FILE = os.path.join(SECRETS_DIR, 'backup_cookies.pkl')

_error_messages = []

_error_messages = []

def log_to_file(message, level='INFO'):
    """Simple logging function that writes to file with timestamp and collects errors."""
    global _error_messages
    try:
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        log_entry = f"{timestamp} - {level} - {message}\n"
        with open(LOGFILE, 'a', encoding='utf-8') as f:
            f.write(log_entry)
        if level == 'ERROR':
            print(f"ERROR: {message}")
            _error_messages.append(f"{timestamp} - {message}")
    except Exception:
        pass  # Fail silently if logging fails

def get_error_messages():
    """Returns the list of captured error messages."""
    return _error_messages

def clear_error_messages():
    """Clears the list of captured error messages."""
    global _error_messages
    _error_messages = []

def get_error_messages():
    """Returns the list of captured error messages."""
    return _error_messages

def clear_error_messages():
    """Clears the list of captured error messages."""
    global _error_messages
    _error_messages = []

def parse_assignment_date(date_str, is_missing=False):
    """Parse various date formats and return standardized date object or None
    
    Args:
        date_str: The date string to parse
        is_missing: True if this is from missing assignments (interpret weekdays as past), 
                   False if from assigned assignments (interpret weekdays as future)
    """
    if not date_str or date_str in ['', 'Unknown', 'No due date']:
        return None
    
    # Skip non-date strings
    if any(keyword in date_str.lower() for keyword in ['posted', 'no due date', 'unknown']):
        return None
    
    try:
        # Handle relative dates
        today = datetime.now()
        clean_date = date_str.strip()
        
        if 'today' in clean_date.lower():
            return today.strftime('%Y-%m-%d')
        elif 'yesterday' in clean_date.lower():
            return (today - timedelta(days=1)).strftime('%Y-%m-%d')
        elif 'tomorrow' in clean_date.lower():
            return (today + timedelta(days=1)).strftime('%Y-%m-%d')
        
        # Handle weekday names ONLY if no actual date is specified
        # Only apply weekday logic if the string contains ONLY a weekday (plus optional time)
        weekdays = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday']
        clean_lower = clean_date.lower()
        
        # Check if this is a weekday-only date (no month names, no numbers that look like dates)
        has_month = any(month in clean_lower for month in [
            'january', 'february', 'march', 'april', 'may', 'june',
            'july', 'august', 'september', 'october', 'november', 'december',
            'jan', 'feb', 'mar', 'apr', 'may', 'jun',
            'jul', 'aug', 'sep', 'oct', 'nov', 'dec'
        ])
        
        # Check for date-like numbers (day/month/year patterns)
        import re
        has_date_numbers = bool(re.search(r'\b\d{1,2}[/\-]\d{1,2}|\b\d{1,2}\s*(st|nd|rd|th)?\s*,|\b\d{4}\b', clean_lower))
        
        # Only apply weekday logic if it's just a weekday (possibly with time)
        if not has_month and not has_date_numbers:
            for weekday in weekdays:
                if weekday in clean_lower:
                    # Find the target weekday
                    target_weekday = weekdays.index(weekday)
                    current_weekday = today.weekday()  # 0=Monday, 6=Sunday
                    
                    if is_missing:
                        # For missing assignments, find the most recent occurrence of this weekday
                        days_ago = (current_weekday - target_weekday) % 7
                        if days_ago == 0:  # If it's the same weekday, assume it was last week
                            days_ago = 7
                        target_date = today - timedelta(days=days_ago)
                    else:
                        # For assigned assignments, find the next occurrence of this weekday
                        days_ahead = (target_weekday - current_weekday) % 7
                        if days_ahead == 0:  # If it's the same weekday, assume it's next week
                            days_ahead = 7
                        target_date = today + timedelta(days=days_ahead)
                    
                    return target_date.strftime('%Y-%m-%d')
        
        # Use dateutil to parse the date
        parsed_date = dateutil_parser.parse(clean_date, default=today)
        return parsed_date.strftime('%Y-%m-%d')
        
    except (ValueError, TypeError) as e:
        log_to_file(f"Could not parse date '{date_str}': {e}", 'WARNING')
        return date_str  # Return original if parsing fails

def create_assignment_object(name, class_name, due_date, due_date_parsed, url, description='', max_points=0):
    """Create standardized assignment object"""
    return {
        'name': name,
        'class': class_name,
        'due_date': due_date,
        'due_date_parsed': due_date_parsed,
        'url': url,
        'description': description,
        'max_points': max_points
    }

def save_json_data(data, filename):
    """Save data to JSON file with error handling"""
    try:
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        log_to_file(f"Data saved to {filename}")
        return True
    except Exception as e:
        log_to_file(f"Failed to save data to {filename}: {e}", 'ERROR')
        return False

def load_json_data(filename):
    """Load data from JSON file with error handling"""
    try:
        if not os.path.exists(filename):
            log_to_file(f"File {filename} does not exist", 'WARNING')
            return None
            
        with open(filename, 'r', encoding='utf-8') as f:
            data = json.load(f)
        log_to_file(f"Data loaded from {filename}")
        return data
    except Exception as e:
        log_to_file(f"Failed to load data from {filename}: {e}", 'ERROR')
        return None

File: scrapers/assignment-utils.js
const fs = require('fs').promises;
const path = require('path');
const { DATA_DIR } = require('config/constants');

// File paths
const ASSIGNMENTS_FILE = path.join(DATA_DIR, 'all_assignments.json');

const WEEKDAYS = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'];
const MONTH_NAMES = [
  'january', 'february', 'march', 'april', 'may', 'june',
  'july', 'august', 'september', 'october', 'november', 'december',
  'jan', 'feb', 'mar', 'apr', 'may', 'jun',
  'jul', 'aug', 'sep', 'oct', 'nov', 'dec'
];

function parseDueDate(dateStr) {
  if (!dateStr || typeof dateStr !== 'string') return null;
  
  const lower = dateStr.toLowerCase().trim();
  
  // Skip non-date strings
  if (['posted', 'no due date', 'unknown'].some(keyword => lower.includes(keyword))) {
    return null;
  }
  
  const today = new Date();
  const currentYear = today.getFullYear();
  
  // Handle relative dates
  if (lower.includes('today')) {
    return formatDateTimeString(today);
  }
  if (lower.includes('yesterday')) {
    const yesterday = new Date(today);
    yesterday.setDate(today.getDate() - 1);
    return formatDateTimeString(yesterday);
  }
  if (lower.includes('tomorrow')) {
    const tomorrow = new Date(today);
    tomorrow.setDate(today.getDate() + 1);
    return formatDateTimeString(tomorrow);
  }
  
  // Check if this is weekday-only (no month names or date numbers)
  const hasMonth = MONTH_NAMES.some(month => lower.includes(month));
  const hasDateNumbers = /\b\d{1,2}[/\-]\d{1,2}|\b\d{1,2}\s*(st|nd|rd|th)?\s*,|\b\d{4}\b/.test(lower);
  
  if (!hasMonth && !hasDateNumbers) {
    const weekdayIndex = WEEKDAYS.findIndex(day => lower.includes(day));
    if (weekdayIndex !== -1) {
      const currentWeekday = today.getDay() === 0 ? 6 : today.getDay() - 1; // Convert to Monday=0
      let daysToAdd = weekdayIndex - currentWeekday;
      if (daysToAdd <= 0) daysToAdd += 7; // Next occurrence
      
      const targetDate = new Date(today);
      targetDate.setDate(today.getDate() + daysToAdd);
      return formatDateTimeString(targetDate);
    }
  }
  
  // Handle time-only strings (assume today)
  const timeOnlyPattern = /^\s*\d{1,2}:\d{2}\s*(AM|PM)?\s*$/i;
  if (timeOnlyPattern.test(dateStr)) {
    const timeStr = dateStr.trim();
    const todayDateStr = today.toISOString().split('T')[0]; // YYYY-MM-DD
    const fullDateStr = `${todayDateStr} ${timeStr}`;
    const parsed = new Date(fullDateStr);
    if (!isNaN(parsed.getTime())) {
      return formatDateTimeString(parsed);
    }
  }
  
  // Attempt standard date parsing with year assumption
  try {
    let parsed = new Date(dateStr);
    
    // If the parsed date has an obviously wrong year (like 2001), fix it
    if (!isNaN(parsed.getTime())) {
      if (parsed.getFullYear() < 2020 || parsed.getFullYear() > currentYear + 1) {
        // Assume current year or next year
        const month = parsed.getMonth();
        const date = parsed.getDate();
        const hours = parsed.getHours();
        const minutes = parsed.getMinutes();
        
        // Smart year determination
        let year = currentYear;
        
        // Create a test date with current year
        const testDate = new Date(currentYear, month, date, hours, minutes);
        const now = new Date();
        
        // If the assignment date is more than 6 months in the past, 
        // it's likely meant for next year
        const monthsDiff = (now.getFullYear() - testDate.getFullYear()) * 12 + 
                          (now.getMonth() - testDate.getMonth());
        
        if (monthsDiff > 6) {
          year = currentYear + 1;
        }
        // If the assignment date is more than 6 months in the future,
        // it's likely meant for last year (missing assignment)
        else if (monthsDiff < -6) {
          year = currentYear - 1;
        }
        
        parsed = new Date(year, month, date, hours, minutes);
      }
      return formatDateTimeString(parsed);
    }
  } catch (e) {
    // Continue to fallback
  }
  
  return null; // Return null if parsing fails completely
}

function formatDateString(date) {
  return date.toISOString().split('T')[0]; // YYYY-MM-DD format
}

function formatDateTimeString(date) {
  // Return ISO string with time: YYYY-MM-DDTHH:MM:SS.sssZ
  return date.toISOString();
}

/**
 * Cleans up course names by removing room numbers, section numbers, etc.
 * Removes everything from the first word containing a digit onwards.
 * @param {string} courseName - The original course name
 * @returns {string} - The cleaned course name
 */
function cleanCourseName(courseName) {
  if (!courseName) return courseName;
  
  const words = courseName.trim().split(/\s+/);
  const cleanWords = [];
  
  for (const word of words) {
    // Check if word contains any digit
    if (/\d/.test(word)) {
      break; // Stop at first word with a digit
    }
    cleanWords.push(word);
  }
  
  // If we'd have an empty string, return the original
  return cleanWords.length > 0 ? cleanWords.join(' ') : courseName;
}

function createAssignmentObject(name, className, dueDate, url, description = '', maxPoints = 0) {
  // Parse the due date for internal use
  const parsed = parseDueDate(dueDate);

  // Prepare a user-friendly display value for the due date.
  // If we have a parsed ISO timestamp, convert to a local date string
  // (no time) to avoid showing UTC offsets like "T04:00:00.000Z" in the UI.
  let displayDue = '';
  if (parsed) {
    const dt = new Date(parsed);
    if (!isNaN(dt.getTime())) {
      // Format like: "Oct 22, 2025"
      displayDue = dt.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
    } else {
      displayDue = dueDate || '';
    }
  } else {
    // If there's no parsed date, keep original string or mark as 'No due date'
    displayDue = dueDate && typeof dueDate === 'string' && dueDate.trim() ? dueDate.trim() : '';
  }

  return {
    name,
    class: cleanCourseName(className),
    // Human-friendly date for display (no timezone artifacts)
    due_date: displayDue || 'No due date',
    // Keep an ISO timestamp (or null) for comparisons and filtering
    due_date_parsed: parsed,
    url,
    description,
    max_points: maxPoints
  };
}

async function saveAssignments(assignments, filename = ASSIGNMENTS_FILE) {
  try {
    await fs.writeFile(filename, JSON.stringify(assignments, null, 2), 'utf8');
    return true;
  } catch (error) {
    console.error(`Failed to save assignments to ${filename}:`, error);
    return false;
  }
}

async function loadAssignments(filename = ASSIGNMENTS_FILE) {
  try {
    const data = await fs.readFile(filename, 'utf8');
    return JSON.parse(data);
  } catch (error) {
    if (error.code === 'ENOENT') return [];
    console.error(`Failed to load assignments from ${filename}:`, error);
    return null;
  }
}

module.exports = {
  parseDueDate,
  createAssignmentObject,
  cleanCourseName,
  saveAssignments,
  loadAssignments,
  DATA_DIR,
  ASSIGNMENTS_FILE
};